### PR TITLE
Delete long-deprecated functions

### PR DESCRIPTION
**Pull request by colemanw**
Overview
----------------------------------------
Deletes old functions that have been emitting noisy deprecations for > 3 years.

Comments
----------------------------------------
This is not exhaustive but @eileenmcnaughton doesn't like big PRs so, more later.

### CODE REVIEW

File: CRM/Contact/BAO/Contact.php [hunk 1]
```diff
@@ -1147,36 +1147,6 @@ public static function processImage() {
...
-   * @param bool $restore
-   *   True to set the is_delete = 1 else false to restore deleted contact,
-   *   i.e. is_delete = 0
-   *
-   * @deprecated
-   *
-   * @return bool
-   * @throws \CRM_Core_Exception
-   */
-  public static function contactTrashRestore($contact, $restore = FALSE) {
```
Comment (eileenmcnaughton) [inline CRM/Contact/BAO/Contact.php:1164]: image

File: CRM/Contact/BAO/Contact/Location.php [hunk 2]
```diff
@@ -59,44 +59,6 @@ public static function getEmailDetails($id, $isPrimary = TRUE, $locationTypeID =
...
-   * Get the sms number and display name of a contact.
-   *
-   * @param int $id
-   *   Id of the contact.
-   * @param string|null $type
-   *
-   * @return array
-   *   tuple of display_name and sms if found, or (null,null)
-   */
-  public static function getPhoneDetails($id, $type = NULL) {
```
Comment (eileenmcnaughton) [inline CRM/Contact/BAO/Contact/Location.php:73]: image

File: CRM/Contact/BAO/Contact.php [hunk 3]
```diff
@@ -1147,36 +1147,6 @@ public static function processImage() {
...
-      CRM_Core_Error::deprecatedFunctionWarning('Use contact.create to restore - this does nothing much');
-      // @todo deprecate calling contactDelete with the intention to restore.
-      $updateParams = [
-        'id' => $contact->id,
-        'is_deleted' => FALSE,
-      ];
-      self::create($updateParams);
-      return TRUE;
-    }
-    return self::contactTrash($contact);
```
Comment (eileenmcnaughton) [inline CRM/Contact/BAO/Contact.php:1177]: Not the last use of ^^ so no follow on needed there

File: CRM/Core/DAO.php [hunk 4]
```diff
@@ -1080,25 +1080,6 @@ public static function objectExists($value, $daoName, $daoID, $fieldName = 'name
...
-   *
-   * @param string $tableName
-   * @param string $columnName
-   * @param bool $i18nRewrite
-   *   Whether to rewrite the query on multilingual setups.
-   *
-   * @return bool
-   *   true if exists, else false
-   */
-  public static function checkFieldExists($tableName, $columnName, $i18nRewrite = TRUE) {
```
Comment (eileenmcnaughton) [inline CRM/Core/DAO.php:1097]: image

File: CRM/Core/OptionGroup.php [hunk 5]
```diff
@@ -328,94 +328,6 @@ public static function lookupValues(&$params, $names, $flip = FALSE) {
...
-  /**
-   * @deprecated - use CRM_Core_PseudoConstant::getLabel
-   *
-   * @param string $groupName
-   * @param $value
-   * @param bool $onlyActiveValue
-   *
-   * @return null
-   */
-  public static function getLabel($groupName, $value, $onlyActiveValue = TRUE) {
```
Comment (eileenmcnaughton) [inline CRM/Core/OptionGroup.php:340]: image

File: CRM/Core/OptionGroup.php [hunk 6]
```diff
@@ -328,94 +328,6 @@ public static function lookupValues(&$params, $names, $flip = FALSE) {
...
-    $label,
-    $labelField = 'label',
-    $labelType = 'String',
-    $valueField = 'value'
-  ) {
-    if (empty($label)) {
-      return NULL;
-    }
-
-    CRM_Core_Error::deprecatedFunctionWarning('CRM_Core_PseudoConstant::getKey');
```
Comment (eileenmcnaughton) [inline CRM/Core/OptionGroup.php:395]: image

File: api/v3/utils.php [hunk 7]
```diff
@@ -457,25 +457,6 @@ function _civicrm_api3_store_values(array $fields, array $params, &$values): boo
...
 
-/**
- * Returns field names of the given entity fields.
- *
- * @deprecated
- * @param array $fields
- *   Fields array to retrieve the field names for.
- * @return array
- */
-function _civicrm_api3_field_names($fields) {
```
Comment (eileenmcnaughton) [inline api/v3/utils.php:468]: image

**Comment by colemanw** [comment]
Thanks for putting up with all this @eileenmcnaughton

**Comment by eileenmcnaughton** [comment]
@colemanw Yeah - it's great to get rid of them.

**Comment by jensschuppe** [comment]
While supporting progress and getting rid of old code, and knowing that what CiviCRM's "public API" actually is is quite vague of a concept, I just want to say that removing deprecated code (as old as it may be) is not following semver, which would require a new major version for backwards-incompatible API changes. This makes extensions relying on older code, which was then best-practice, be incompatible with a new minor version of CiviCRM, as e.g. systopia/de.systopia.donrec#164 shows.

**Comment by jensschuppe** [comment]
I see the dilemma, @colemanw - maybe it's just not clear enough to extension developers what a *noisy deprecation* actually means. The example I mentioned is old code, sure, but it used to work fine and nobody felt urged to do/fund a refactoring (though it's just a small change I admit).

Would it be feasible to define a time frame for such deprecations before the code actually gets deleted with reporting to the status report (*You have deprecated code running that is likely to be deleted within the next release/x months/years*)? Do you (core devs) think it's worth discussing that in a GitLab issue, i. e. do you think there'll be more deprecations and removals before CiviCRM 6.0.0?

**Comment by mattwire** [comment]
The most sensible thing would probably be an automated script that puts them into a list and ideally detects which version they got removed. Most have signatures that can be picked up (eg. deprecated tags in code). Then that list could be referred to either manually or automatically - eg. would be great to have a tool (or function in civix) that scans your extension for deprecated code by version.

**Comment by eileenmcnaughton** [comment]
I think we need to be clearer about what is supported regarding the api - we did put quite a lot of work into communicating to people that if they use the public api / hooks then we will manage compatibility & anything else is user-beware - but I feel like we haven't been pushing that message as clearly recently

**Comment by totten** [comment]
> Would it be feasible to define a time frame for such deprecations before the code actually gets deleted with reporting to the status report 

It might straight-forward to improve the current runtime warning:

```diff
-function deprecatedFunctionWarning(string $newMethod, ?string $oldMethod = NULL)
+function deprecatedFunctionWarning(string $newMethod, ?string $oldMethod = NULL, 
+    ?string $since = NULL, ?string $until = NULL)
```

Example:

```php
deprecatedFunctionWarning('doStuff', 'doStuffBetter', '2019-02', '2020-05');
```
```
Warning: doStuff() deprecated in favor of doStuffBetter() as of 2019-02. 
It may be removed as early as 2020-05 or v6.0 (whichever is earlier)."
```

The above is fairly quick/simple. It also lets us start recording some information about the timeline for each method.

Of course, I also like the idea of some static-analysis. 

----

If you can excuse a color-code analogy, the problem is that Civi's PHP space includes:

* 🟢  _Green Symbols_: Things which are OK to use. They're well-supported / widely-used / documented / etc.  Examples: `CRM_Core_Page`, `Civi::settings()`, `civicrm_api()`, `Civi\Api4\X::y()`
* 🔴 _Red Symbols_: Things which should be actively avoided. They're formally deprecated or internal or experimental. Example: `contactTrashRestore()`
* 🟡 _Yellow Symbols_: Murky things. There is no agreement or formal designation. The vast majority of functions have no designation. These are subjective. Probably, some should be supported+documented. Probably, some should be disclaimed or removed. But the main thing is that this list is long and ambiguous.

As someone who writes or runs extensions, one should be able to get a report of the 🔴  and 🟡 symbols.

**Comment by eileenmcnaughton** [comment]
Yeah - I think *supported* is only the green things For the yellow things we try to be helpful but do not commit to supporting them IMHO - I think our efforts to be helpful have reduced the clarity of our efforts to make it clear they are not actually supported

Re the deprecated warning - there is actually a drupal doc block standard for stating when something was deprecated & when it will be removed (not sure if it was just drupal - it was one I saw but had to disable when testing @michaelmcandrew's coder)

**Comment by totten** [comment]
Yeah, I definitely like the idea of putting that information in a standardized docblock. For reference, here are a few interpretations of `@deprecated`:

* https://docs.phpdoc.org/latest/guide/references/phpdoc/tags/deprecated.html, https://github.com/php-fig/fig-standards/blob/master/proposed/phpdoc-tags.md#54-deprecated 
    ```php
    /**
     * @deprecated [<"Semantic Version">] [<description>]
     * @deprecated 1.0.0 No longer used by internal code and not recommended.
    ```
* https://www.drupal.org/about/core/policies/core-change-policies/drupal-deprecation-policy
    ```php
    /**
     * @deprecated in drupal:8.4.0 and is removed from drupal:9.0.0. Use
     * \Drupal\Core\Entity\EntityViewBuilder instead.
    ```
* https://wiki.php.net/rfc/deprecated_attribute (*open/unresolved RFC*)
    ```php
    #[Deprecated("use test3() instead")]
    function test2() {}
    ```

The Drupal approach of standardizing the blurb sounds nice+simple.

I like that there are a few tools which will pick up on `@deprecated` and `@internal` (PHPStorm, Psalm, PHPStan... though maybe that's another plugin...). OTOH, the canonical rules for Civi are in `phpcs`, and AFAICS it will only flag [deprecated callers for php stdlib](https://github.com/squizlabs/PHP_CodeSniffer/blob/b245bb315dae6dc7347681e4f5a55dd7bdfcd045/src/Standards/Generic/Sniffs/PHP/DeprecatedFunctionsSniff.php) -- it the lacks kind of autoloading and type-rules that you need to detect those. 

Also, I also don't think there's any existing ruleset to flag ambiguous 🟡  references. Flagging those would probably require (a) some kind of plugin for psalm or phpstan and/or (b) some kind of script on top of `nikic/php-parser`.

**Comment by eileenmcnaughton** [comment]
Yeah - I don't think it's on us to create a tool for people to find deprecated code calls - but we could improve our commenting - I would stick with the drupal standard since that is our current coding standard - the updated drupal coder enforces the drupal standard

**Comment by bjendres** [comment]
> would be great to have a tool (or function in civix) that scans your extension for deprecated code by version.

See https://github.com/totten/civix/issues/296

**Pull request merged**